### PR TITLE
Add FiniteDifferences based scalar rule tests

### LINKED ISSUE #76: Add FiniteDifferences based tests to some scalar functions
As found for `hypot` 
https://github.com/JuliaDiff/ChainRules.jl/pull/74
we are not testing some of our simpler scalar rules with FiniteDifferences.
We have easy helpers for how to do that in `test/test_utils.jl`
but we just are not using them

### PR DESCRIPTION
**Pull request by oxinabox**
Closes #76
and the mistakes found in #86 for incorrect acshc asec acsc, asecd acscd rules

### CODE REVIEW

File: test/test_util.jl [hunk 1]
```diff
@@ -1,13 +1,43 @@
...
+    @testset "$f at $x, $(nameof(rule))" for rule in (rrule, frule)
+        res = rule(f, x)
+        @test res !== nothing  # Check the rule was defined
+        fx, ∂x = res
+        @test fx == f(x)  # Check we still get the normal value, right
+
+        # Check that we get the derivative right:
+        @test isapprox(
+            ∂x(1), fdm(f, x);
+            rtol=rtol, atol=atol, kwargs...
```
Comment (willtebbutt) [inline test/test_util.jl]: Do we need to have `atol` and `rtol` outside of `kwargs`? If we don't set the defaults to `1e-9`, do we start to get errors?

**Comment by oxinabox** [comment]
> Assuming that we now have 100% coverage of the scalars rules through the use of test_scalar, I'm happy for this to be merged.

We don't yet.
I''m going to add a few more commits to get close to that

File: test/test_util.jl (hunk 1 above)
Comment (oxinabox) [reply test/test_util.jl]: idk, I was copying the `test_frule` and `test_rrule` setup

**Comment by nickrobinson251** [comment]
This is great

**Comment by oxinabox** [comment]
we got the derviative of `inv(x)` wrong...

**Comment by willtebbutt** [comment]
> we got the derviative of inv(x) wrong...

lol

**Comment by oxinabox** [comment]
OK.
Done and ready for review

File: test/rulesets/Base/base.jl [hunk 2]
```diff
@@ -1,66 +1,55 @@
...
+            test_scalar(cospi, x)
         end
-        @testset "Hyperbolic" for x = (Float64(π), Complex(π, π/2))
-            test_scalar(sinh, cosh, x)
-            test_scalar(cosh, sinh, x)
-            test_scalar(tanh, x -> sech(x)^2, x)
-            test_scalar(sech, x -> -tanh(x) * sech(x), x)
-            test_scalar(csch, x -> -coth(x) * csch(x), x)
-            test_scalar(coth, x -> -csch(x)^2, x)
+        @testset "Hyperbolic" for x = (Float64(π)-0.01, Complex(π-0.01, π/2))
```
Comment (nickrobinson251) [inline test/rulesets/Base/base.jl]: out of interest, why this change?

File: test/rulesets/Base/base.jl [hunk 3]
```diff
@@ -83,40 +72,63 @@ end
...
         end
-    end
-    @testset "Misc. Tests" begin
-        @testset "*(x, y)" begin
-            x, y = rand(3, 2), rand(2, 5)
-            z, (dx, dy) = rrule(*, x, y)
+    end  # Trig
 
-            @test z == x * y
+    @testset "math" begin
```
Comment (nickrobinson251) [inline test/rulesets/Base/base.jl]: i like that the rest isn't "math" 😆

File: test/rulesets/Base/base.jl [hunk 4]
```diff
@@ -83,40 +72,63 @@ end
...
 
-            @test dx(z̄) == extern(accumulate(zeros(3, 2), dx, z̄))
-            @test dy(z̄) == extern(accumulate(zeros(2, 5), dy, z̄))
+            test_scalar(exp, x)
+            test_scalar(exp2, x)
+            test_scalar(exp10, x)
 
-            test_accumulation(rand(3, 2), dx, z̄, z̄ * y')
-            test_accumulation(rand(2, 5), dy, z̄, x' * z̄)
+            x isa Real && test_scalar(cbrt, x)
```
Comment (nickrobinson251) [inline test/rulesets/Base/base.jl]: TIL there is a `cbrt` function

**Comment by nickrobinson251** [comment]
we should re-run CI for Julia 1.0?

File: test/rulesets/Base/base.jl (hunk 2 above)
Comment (oxinabox) [reply test/rulesets/Base/base.jl]: Movew it away from the boundries where it isn't defined and we thus can;t use `central_fdm`

Comment (nickrobinson251) [reply test/rulesets/Base/base.jl]: seems fair enough to me 👍

**Comment by simeonschaub** [comment]
@nickrobinson251 It seems that Travis needs to be reconfigured to not abort if tests take longer than 10 minutes

**Comment by nickrobinson251** [comment]
> @nickrobinson251 It seems that Travis needs to be reconfigured to not abort if tests take longer than 10 minutes

Yes, we should extend that time limit - I don't think I have permissions to do that, but hopefully @oxinabox does? It's possible that allowing more time won't solve this case, if e.g. SpecialFunctions is just hanging for some reason on 1.1 (don't know if that's what's happening or not)

**Comment by oxinabox** [comment]
It is not possible to increase the time without log output in Travis.
Plus in the versions that pass the total time is not much more than 10 minutes.
(and sometimes is much less)
So it can't be that SpecialFunctions is suposed to need 11minutes to build or something

**Comment by simeonschaub** [comment]
Is the solution just to insert random print statements?

**Comment by oxinabox** [comment]
> Is the solution just to insert random print statements?

Apparently.

I wonder if we should just get rid of the top-level TestSet so it prints each testset as it completes.
